WebUSB error "Cannot read property 'replace' of undefined"

microbit-sam opened the pull request:
Check that `err.message` exists before using it

---

May not be necessary, updated translation handling can probably remove the possibility of this happening

microbit-carlos commented:
Have you been able to replicate the original issue? Or do you know the type of the object that gets to this point?
Basically we are getting something that is not an `Error` object, and I think we need to understand what that is and why that happens, as this might be a symptom of a different problem.

microbit-sam commented:
No I've not managed to discover that yet, almost all of the calls to `webusbErrorHandler` are from `.catch()` events so should have them

The only call that isn't from a `catch()` is 

```
  var error = {"name": "device-disconnected", "message": config["translate"]["webusb"]["err"]["device-disconnected"]};
        webusbErrorHandler(error);
```

I'm checking translations now, but I would have thought `message` would still have been defined even if there was a missing translation

microbit-carlos commented:
That sounds like a good candidate as the original issue was found when the editor was in Spanish.

microbit-sam commented:
If so, shall I move the check to:

```
  var error = {"name": "device-disconnected", "message": (config["translate"]["webusb"]["err"]["device-disconnected"] ? config["translate"]["webusb"]["err"]["device-disconnected"] : "missing-translation" };
        webusbErrorHandler(error);
```

microbit-sam commented:
Actually I've just realised, the metrics will have seperate URLs for errors in different languages.. that's not ideal

microbit-carlos commented:
I think we need a better way to deal with missing translations. Likely to create a "getter" that returns the english version if the string is not found.

microbit-sam commented:
> I think we need a better way to deal with missing translations. Likely to create a "getter" that returns the english version if the string is not found.

Yep that sounds like a good idea

microbit-carlos commented:
Saw this error again while testing PR #270:
![image](https://user-images.githubusercontent.com/29712657/66942080-291f4880-f040-11e9-98db-87225e576306.png)

So it looks like it's not related to the translations after all.

microbit-sam commented:
Interesting, do you still have that window open? I think errors should be logged when webusbErrorHandler is called

microbit-sam commented:
I've restructured webusbErrorHandler, and it feels better to me, but you may have other thoughts Carlos!

- `replace()` should always be called on a string now
- Moved errors such as `Error: Bad response for 140 -> 5` to the console, and replaced the user facing text with a generic WebUSB Error
- Added console.trace() to the error handler

microbit-carlos commented on python-main.js:
```diff
@@ -1266,63 +1266,116 @@ function web_editor(config) {
+                if (usePartialFlashing) {
+                    errorDescription += '<br>' + config["translate"]["webusb"]["err"]["partial-flashing-disable"];
+                } 
+                break;
```
We are not doing anything with the `err` string and just using a generic error title, which is passed to the metrics.
So we will never be able to capture this error in the metrics, and the user sending us screenshots will have generic text that won't give us any indication either.

Can we surface this string in a way that will help us identify issues in this case?

microbit-sam replied:
Done

microbit-carlos commented on python-main.js:
```diff
@@ -1266,63 +1266,116 @@ function web_editor(config) {
-            } 
+        // Display error handler modal
+        $("#flashing-overlay-container").css("display", "flex");
+        $("#flashing-info").addClass('hidden');
```
If an exception is thrown before this line, will the flashing modal be "stuck"?

microbit-carlos commented on python-main.js:
```diff
@@ -1266,63 +1266,116 @@ function web_editor(config) {
+                    if (usePartialFlashing) {
+                        errorDescription += '<br>' + config["translate"]["webusb"]["err"]["partial-flashing-disable"];
+                    } 
+                }
```
As mentioned in a previous comment, we are swallowing error types, I think we should probably show the user.
For instance, if the code is too large, a microbit-fs error ends up here that is never shown to the user.
`Error: Oops! Could not load the code into the hex file, Not enough space for the main.py file.`

microbit-sam replied:
How about something along these lines?

![Screenshot 2019-10-22 at 13 29 27](https://user-images.githubusercontent.com/32453267/67281569-0376c100-f4d0-11e9-8e8a-f22ed998aa26.png)

microbit-carlos replied:
Maybe with a much lighter background:
![image](https://user-images.githubusercontent.com/29712657/67282609-15edfc00-f4ca-11e9-844a-32a83d860f50.png)

```css
{
    margin: 16px 0px;
    padding: 12px;
    border-radius: 8px;
    background-color: #fafafa;
}
```

microbit-sam replied:
👍

microbit-carlos commented on python-main.js:
```diff
@@ -1265,64 +1265,118 @@ function web_editor(config) {
+                    errorType = "reconnect-microbit";
+                    errorTitle = "WebUSB Error";
+                    errorDescription = config["translate"]["webusb"]["err"][errorType];
+                    if (usePartialFlashing) {
```
Can we also add a check for the experimental flag here and in the other instances of this part of the error message?

As only the beta version of the editor can disable Quick Flash, the message "If the errors persist, try disabling Quick Flash in the beta options." should not appear in the primary editor.

microbit-sam replied:
👍
